sox/transform.py

```diff
@@ -454,6 +461,102 @@ def build(self, input_filepath, output_filepath, extra_args=None,
             else:
                 return True
 
+    def build_array(self, input_array, sample_rate_in, sample_rate_out=None,
+                    channels_in=None, channels_out=None, bits_in=None, 
+                    bits_out=None, encoding_out=None, extra_args=None,
+                    return_output=True):
+        '''Builds an output numpy array by executing the current set of 
+        commands. Returns a tuple containing the status, the output array,
+        and the stderr: (status, output_array, stderr).
+
+        Parameters
+        ----------
+        input_array : np.ndarray
+            Input audio as a numpy array - (samples, channels).
+        sample_rate_in : int
+            Sample rate of audio. Since the data is given as a numpy 
+            array, we need this at build to set the input format correctly.
+        sample_rate_out : int, default=None
+            Expected sample rate of output audio. If None, this is set to
+            sample_rate_in.
+        channels_in : int, default=None
+            Number of channels in input. If None, defaults to number of channels
+            in input_array.
+        channels_out : int, default=None
+            Expected number of channels in output audio. If None, this is set 
+            to channels_in.
+        bits_in : int, default=None
+            Bitdepth of input audio. If None, inferred at runtime by sox. 
+        bits_out : int, default=None
+            Expected bitdepth of output audio. If None, inferred by sox.
+        encoding_out : str or None, default=None
+            str determining expected encoding of output audio. Must be one of
+            's8', 's16', 'f32', 'f64'.
+        extra_args : list or None, default=None
+            If a list is given, these additional arguments are passed to SoX
+            at the end of the list of effects.
+            Don't use this argument unless you know exactly what you're doing!
+        '''
+        output_filepath = '-'
```

Comment:
Thinking out loud... does it also make sense to support array in --> file out, and file in --> array out? My thinking is yes, in which case maybe all or this function should be part of `build`? What do you think @pseeth ?

Reply:
Totally possible. Do you have a suggested function signature for `build`? It seems like a lot of the arguments will become optional, or do type-checking on input/output (in which case the arguments will be renamed). Then we'll have some sort of merge between the `build_array` and `build` functions, which I think would look reasonable. Let me know your thoughts!

Reply:
This is supported now! I merged `build` and `build_array` to be able to support this. There's some finicky typechecking that needed to happen, and a flag that tells the `core.sox` function whether to decode `out` with `utf-8` but it now works! I expanded the test cases to test `file -> array`, `array -> array`, and `array -> file`.